### DISCUSSION TITLE
Insister sur l'importance de la stack, en particulier du language

### DISCUSSION THREAD

**Discussion by gagarine** [cnumr/best-practices#410] [Bonnes pratiques]
Hello,

La question du langage de programmation et de la stack en général ne me semble pas abordée explicitement. Je m'explique.

Les différences de performances entre langages sont parfois très importantes (rapidité d’exécution) mais aussi d'efficacité (utilisation de ressource), de mon expérience ce n'est pas forcément corrélé de manière linéaire (on peut avoir des langages très rapides, mais qui utilise beaucoup de ressource).

Par exemple Go est rapide, mais utilise beaucoup de ressources (RAM et cycle CPU). Rust est largement moins demandeur en ressources. (on peut faire une analogie avec Firefox et Chrome, les deux navigateurs ont des performances similaires, mais Chrome utiliser beaucoup plus de RAM).

Typiquement, nodejs est aujourd'hui souvent plus lent que PHP tout en demandant davantage de ressource https://phabricator.wikimedia.org/phame/post/view/189/parsoid_in_php_or_there_and_back_again/ 

Je ne sais pas exactement comment formuler ce point, mais il me semble important.

Un dev web est souvent focalisé sur les performances (temps d'affichage). Il teste son site sur une machine performante et avec peu de tabs ouverte. Dans la vraie vie, le site est une tab parmi d'autre et utilisation excecissve de ressource va embourbé la machine cliente et motivé l'utilisateur à changé d'ordinateur... ce qui est le vrai problème.

L'usage d'un ordinateur n'est pas si polluant relativement à sa construction.

**Comment by gagarine** [comment]
Une ressources intéressante https://greenlab.di.uminho.pt/wp-content/uploads/2017/10/sleFinal.pdf (Energy Efficiency across Programming Languages)

**Comment by tbroyer** [comment]
J'ai l'impression qu'on mélange plein de choses là.
Côté client: une des clés est l'optimisation technique. Mais le langage a peu d'importance puisqu'in fine tout est JavaScript (ou WASM). Le framework aura une importance par contre: en quantité de code (React vs Preact) et en performance d'exécution (React vs Solid). La gestion des CSS (Shadow DOM 👍), le lazy loading, etc. Tout ça est il me semble déjà abordé.

Côté serveur, l'impact est infiniment moindre. Pas inexistant, mais négligeable comparé à des choix d'architecture bien plus importants (micro services vs "majestic monolith"). Et le tout dépend énormément du besoin de passage à l'échelle. Et encore une fois je crois que ces aspects sont déjà traités.

**Reply by gagarine** [reply]
Oui, c'est le cas pour le web "classic". Mais lorsque l'on utilise des technologies web pour faire des applications (notamment mobile, CLI ou desktop avec electron) les performances sont désastreuses. Effectivement le guide ce focalise sur le web et non le "digital" ou internet en général. Mais la culture "front-end" à influencé le développement de logiciel en général. La temptation de mettre du JS partout se fait souvent en méconnaissance des pertes en performances.
L'architecture à un impact important, mais à architecture égal le choix du language modifie les performance de manière substantielles.

**Comment by WebMamba** [comment]
Désoler mais je ne suis pas d'accord avec ce point. Je pense qu'il y a plus d'intérêt à ce que les développeur utilise un language qu'ils maitrisent pour ensuite bien optimiser. Que partir sur language plus bas niveau, et ensuite galérer sur le développement.
